### PR TITLE
[SPARK-42522][CONNECT] Fix DataFrameWriterV2 to find the default source

### PR DESCRIPTION
**Pull request by ueshin**
### What changes were proposed in this pull request?

Fixes `DataFrameWriterV2` to find the default source.

### Why are the changes needed?

Currently `DataFrameWriterV2` in Spark Connect doesn't work without the provider with a weird error:

For example:

```py
df.writeTo("test_table").create()
```

```
pyspark.errors.exceptions.connect.SparkConnectGrpcException: (org.apache.spark.SparkClassNotFoundException) [DATA_SOURCE_NOT_FOUND] Failed to find the data source: . Please find packages at `https://spark.apache.org/third-party-projects.html`.
```

### Does this PR introduce _any_ user-facing change?

Users will be able to use `DataFrameWriterV2` without provider as same as PySpark.

### How was this patch tested?

Added some tests.

### CODE REVIEW

**Review by HyukjinKwon — APPROVED** [review]

**Comment by ueshin** [comment]
The tests fail with the error:

```
pyspark.errors.exceptions.connect.SparkConnectGrpcException: <_MultiThreadedRendezvous of RPC that terminated with:
	status = StatusCode.UNAVAILABLE
	details = "failed to connect to all addresses; last error: UNKNOWN: ipv4:127.0.0.1:44883: Failed to connect to remote host: Connection refused"
	debug_error_string = "UNKNOWN:failed to connect to all addresses; last error: UNKNOWN: ipv4:127.0.0.1:44883: Failed to connect to remote host: Connection refused {grpc_status:14, created_time:"2023-02-22T00:04:47.405869565+00:00"}"
>
```

They pass in my local env, though.

cc @HyukjinKwon @zhengruifeng

**Comment by HyukjinKwon** [comment]
let me take a quick look

**Comment by HyukjinKwon** [comment]
I can only reproduce this via running `./python/run-tests --testnames 'pyspark.sql.tests.connect.test_parity_readwriter'` .. cannot reproduce it via IDE too

**Comment by HyukjinKwon** [comment]
okay .. finally found out why .. let me create a PR ..

**Comment by hvanhovell** [comment]
retriggered tests

File: connector/connect/common/src/main/protobuf/spark/connect/commands.proto [hunk 1]
```diff
@@ -128,7 +128,7 @@ message WriteOperationV2 {
 
   // (Optional) A provider for the underlying output data source. Spark's default catalog supports
   // "parquet", "json", etc.
-  string provider = 3;
+  optional string provider = 3;
```
Comment (amaliujia) [inline connector/connect/common/src/main/protobuf/spark/connect/commands.proto:131]: There is no need for `optional`.

Comment (ueshin) [reply connector/connect/common/src/main/protobuf/spark/connect/commands.proto:131]: Without `optional`, protobuf doesn't provide `hasXxx` method and we can't check it's set/unset.

Comment (ueshin) [reply connector/connect/common/src/main/protobuf/spark/connect/commands.proto:131]: We can omit `optional` only when we can see the default value is the same as unset, e.g., empty list/map and unset. Even for empty list/map, if we can't see they are the same, we should use `optional`. See #39960 for example.

In this case, the default value for string `""` is different from unset.

Comment (amaliujia) [reply connector/connect/common/src/main/protobuf/spark/connect/commands.proto:131]: Sorry I actually miss-remember this: the complete  dropping `optional` is still a proposal but not finalized yet. This line is good.

**Comment by amaliujia** [comment]
LGTM

**Comment by hvanhovell** [comment]
merging this

**Pull request closed without merging**